File: src/lib/workflow.ts
import type { Role, Stage } from '@/types'

/**
 * Central workflow metadata — the single source of truth for how an RFQ moves
 * through the pipeline. Stage-mutating logic lives in the rfq store; this file
 * only describes the shape of the state machine (spec §2, §41, §42, §43).
 */

export const STAGE_ORDER: Stage[] = [
  'Draft',
  'Operations Review',
  'Sourcing',
  'Controlling',
  'Approval Pending',
  'Approved',
  'Quotation Generated',
  'Quotation Sent',
]

export const FUNNEL_STAGES: Stage[] = [
  'Draft',
  'Operations Review',
  'Sourcing',
  'Controlling',
  'Approval Pending',
  'Quotation Generated',
]

/** Which role must act while the RFQ sits in a given stage. */
export const STAGE_OWNER: Record<Stage, Role | null> = {
  Draft: 'Sales',
  'Operations Review': 'Operations',
  Sourcing: 'Sourcing',
  Controlling: 'Controlling',
  'Approval Pending': 'Approval Panel',
  Approved: 'Sales',
  'Quotation Generated': 'Sales',
  'Quotation Sent': 'Sales',
  Won: null,
  Lost: null,
  Rejected: null,
}

/** Forward progression when a stage's owner takes the primary action. */
export const NEXT_STAGE: Partial<Record<Stage, Stage>> = {
  Draft: 'Operations Review',
  'Operations Review': 'Sourcing',
  Sourcing: 'Controlling',
  Controlling: 'Approval Pending',
  'Approval Pending': 'Approved',
  Approved: 'Quotation Generated',
  'Quotation Generated': 'Quotation Sent',
}

/** Valid "Send Back" targets per spec §2 / §42. */
export const SEND_BACK_TARGETS: Partial<Record<Stage, Stage[]>> = {
  'Operations Review': ['Draft'],
  Sourcing: ['Operations Review', 'Draft'],
  Controlling: ['Sourcing', 'Draft'],
  'Approval Pending': ['Controlling', 'Sourcing', 'Draft'],
}

export const SEND_BACK_LABELS: Record<Stage, string> = {
  Draft: 'Sales',
  'Operations Review': 'Operations',
  Sourcing: 'Sourcing',
  Controlling: 'Controlling',
  'Approval Pending': 'Approval Panel',
  Approved: 'Sales',
  'Quotation Generated': 'Sales',
  'Quotation Sent': 'Sales',
  Won: 'Sales',
  Lost: 'Sales',
  Rejected: 'Sales',
}

/** Which stages a role is allowed to reject an RFQ from. */
export const REJECTABLE_STAGES: Stage[] = ['Operations Review', 'Sourcing', 'Controlling', 'Approval Pending']

export const REJECTION_REASONS = [
  'Not technically feasible',
  'Pricing not viable',
  'Customer requirements unclear',
  'Vendor unavailable',
  'Duplicate RFQ',
  'Other',
]

export const LOSS_REASONS = ['Price', 'Delivery', 'Competitor', 'Technical', 'Customer Cancelled', 'Other']

export const STAGE_TONE: Record<Stage, 'neutral' | 'teal' | 'purple' | 'orange' | 'blue' | 'green' | 'red' | 'amber'> = {
  Draft: 'neutral',
  'Operations Review': 'teal',
  Sourcing: 'purple',
  Controlling: 'orange',
  'Approval Pending': 'blue',
  Approved: 'green',
  'Quotation Generated': 'blue',
  'Quotation Sent': 'blue',
  Won: 'green',
  Lost: 'red',
  Rejected: 'red',
}

export function canRoleActOnStage(role: Role, stage: Stage): boolean {
  if (role === 'Admin') return true
  return STAGE_OWNER[stage] === role
}

export function isTerminalStage(stage: Stage): boolean {
  return stage === 'Won' || stage === 'Lost' || stage === 'Rejected'
}

File: backend/rfq/workflow.py
"""
Ports src/lib/workflow.ts exactly — the single source of truth for how an RFQ moves
through the pipeline. Previously the frontend was the only thing enforcing this; now
that RFQs are shared backend data, every transition is re-validated here so one
client can't skip a stage no matter what the UI sends.
"""

STAGE_OWNER = {
    "Draft": "Sales",
    "Operations Review": "Operations",
    "Sourcing": "Sourcing",
    "Controlling": "Controlling",
    "Approval Pending": "Approval Panel",
    "Approved": "Sales",
    "Quotation Generated": "Sales",
    "Quotation Sent": "Sales",
    "Won": None,
    "Lost": None,
    "Rejected": None,
}

NEXT_STAGE = {
    "Draft": "Operations Review",
    "Operations Review": "Sourcing",
    "Sourcing": "Controlling",
    "Controlling": "Approval Pending",
    "Approval Pending": "Approved",
    "Approved": "Quotation Generated",
    "Quotation Generated": "Quotation Sent",
}

SEND_BACK_TARGETS = {
    "Operations Review": ["Draft"],
    "Sourcing": ["Operations Review", "Draft"],
    "Controlling": ["Sourcing", "Draft"],
    "Approval Pending": ["Controlling", "Sourcing", "Draft"],
}

REJECTABLE_STAGES = ["Operations Review", "Sourcing", "Controlling", "Approval Pending"]


class WorkflowError(Exception):
    """Raised when a requested transition isn't valid from the RFQ's current stage."""


def require_stage(rfq, expected_stage: str) -> None:
    if rfq.stage != expected_stage:
        raise WorkflowError(f"RFQ is in '{rfq.stage}', not '{expected_stage}' — action not allowed.")


def require_role_can_act(role: str, stage: str) -> None:
    if role == "Admin":
        return
    if STAGE_OWNER.get(stage) != role:
        raise WorkflowError(f"Role '{role}' cannot act on stage '{stage}' (owned by '{STAGE_OWNER.get(stage)}').")


def require_valid_send_back(from_stage: str, target_stage: str) -> None:
    targets = SEND_BACK_TARGETS.get(from_stage, [])
    if target_stage not in targets:
        raise WorkflowError(f"Cannot send back from '{from_stage}' to '{target_stage}' — valid targets are {targets}.")


def require_rejectable(stage: str) -> None:
    if stage not in REJECTABLE_STAGES:
        raise WorkflowError(f"RFQ in stage '{stage}' cannot be rejected.")
